<h1 align=center><font size = 3>Segmenting and Clustering Neighborhoods in city of Toronto</font></h1>

### Introduction

Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. WE will scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas  dataframe. WE will convert addresses like postalcode and borough information into their equivalent latitude and longitude values. Also, will use the Foursquare API to explore neighborhoods in Toronto City. Will use the explore function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters using the k-means clustering algorithm to complete this task. Finally, will use the Folium library to visualize the neighborhoods in Toronto City and their emerging clusters.

<h2 id="DSCW">Downloading And Scraping The Contents Of A Web Page</h2> 


In [1]:
!pip install bs4
#!pip install requests

Import the required modules and functions


In [2]:
from bs4 import BeautifulSoup # this module helps in web scrapping.
import requests  # this module helps us to download a web page
import pandas as pd

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values


#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# import k-means from clustering stage
from sklearn.cluster import KMeans

import numpy as np # library to handle data in a vectorized manner

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

We Download the contents of the web page:

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

We use <code>get</code> to download the contents of the webpage in text format and store in a variable called <code>data</code>:


In [4]:
data  = requests.get(url).text

We create a <code>BeautifulSoup</code> object using the <code>BeautifulSoup</code> constructor


In [5]:
soup = BeautifulSoup(data,"html5lib")  # create a soup object using the variable 'data'

In [6]:
#find a html table in the web page
table = soup.find('table') # in html table is represented by the tag <table>

We Create a dataframe from the table data consist of three columns: PostalCode, Borough, and Neighborhood.

In [7]:
table_contents=[]
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

Print dataframe

In [8]:
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
# Shape of the dataframe
df.shape

(103, 3)

 Read csv file that has the geographical coordinates of each postal code and create a dataframe with csv data

In [10]:
df_ll = pd.read_csv("Geospatial_Coordinates.csv")
df_ll


Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Renmame the coulmn 'Postal Code' of df_ll as same as column name 'PostalCose' of previous dataframe df

In [11]:
df_ll.rename(columns = {"Postal Code": "PostalCode"}, inplace=True)

Merge two dataframe df and df_ll into a new dataframe named df_new

In [12]:
df_new = pd.merge(df, df_ll, on='PostalCode')
df_new['Borough'].unique()

array(['North York', 'Downtown Toronto', "Queen's Park", 'Etobicoke',
       'Scarborough', 'East York', 'York', 'East Toronto', 'West Toronto',
       'East York/East Toronto', 'Central Toronto', 'Mississauga',
       'Downtown Toronto Stn A', 'Etobicoke Northwest',
       'East Toronto Business'], dtype=object)

And make sure that the dataset has all 15 boroughs and 103 neighborhoods.


In [13]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_new['Borough'].unique()),
        df_new.shape[0]
    )
)

The dataframe has 15 boroughs and 103 neighborhoods.


### Explore and cluster the neighborhoods in Toronto.

#### Use geopy library to get the latitude and longitude values of Toronto.


In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.


In [14]:
address = 'Toronto, Canada'
geolocator = Nominatim(user_agent='tr-explorer')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


As we did with all of Toronto, let's visualize only boroughs that contain the word Toronto the neighborhoods in it.


In [15]:
borough_toronto = []
for name in df_new['Borough']:
    if 'Toronto' in name:
        if name in borough_toronto:
            pass
        else:
            borough_toronto.append(name)
print(borough_toronto)
            

['Downtown Toronto', 'East Toronto', 'West Toronto', 'East York/East Toronto', 'Central Toronto', 'Downtown Toronto Stn A', 'East Toronto Business']


#### Convert Borough name 'Downtown Toronto o Stn A' to 'Downtown Toronto'

In [16]:
df_new['Borough'] = df_new['Borough'].replace(['Downtown Toronto Stn A'],'Downtown Toronto')
df_new['Borough'].unique()

array(['North York', 'Downtown Toronto', "Queen's Park", 'Etobicoke',
       'Scarborough', 'East York', 'York', 'East Toronto', 'West Toronto',
       'East York/East Toronto', 'Central Toronto', 'Mississauga',
       'Etobicoke Northwest', 'East Toronto Business'], dtype=object)

#### Create a map of Toronto with neighborhoods superimposed on top.


In [17]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_new['Latitude'], df_new['Longitude'], df_new['Borough'], df_new['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [78]:
toronto_data = df_new[df_new['Borough'].isin(borough_toronto)].reset_index(drop=True)
toronto_data

PostalCode                 Borough  \
0         M5A        Downtown Toronto   
1         M5B        Downtown Toronto   
2         M5C        Downtown Toronto   
3         M4E            East Toronto   
4         M5E        Downtown Toronto   
5         M5G        Downtown Toronto   
6         M6G        Downtown Toronto   
7         M5H        Downtown Toronto   
8         M6H            West Toronto   
9         M4J  East York/East Toronto   
10        M5J        Downtown Toronto   
11        M6J            West Toronto   
12        M4K            East Toronto   
13        M5K        Downtown Toronto   
14        M6K            West Toronto   
15        M4L            East Toronto   
16        M5L        Downtown Toronto   
17        M4M            East Toronto   
18        M4N         Central Toronto   
19        M5N         Central Toronto   
20        M4P         Central Toronto   
21        M5P         Central Toronto   
22        M6P            West Toronto   
23        M4R         Central Toronto   
24        M5R         Central Toronto   
25        M6R            West Toronto   
26        M4S         Central Toronto   
27        M5S        Downtown Toronto   
28        M6S            West Toronto   
29        M4T         Central Toronto   
30        M5T        Downtown Toronto   
31        M4V         Central Toronto   
32        M5V        Downtown Toronto   
33        M4W        Downtown Toronto   
34        M5W        Downtown Toronto   
35        M4X        Downtown Toronto   
36        M5X        Downtown Toronto   
37        M4Y        Downtown Toronto   
38        M7Y   East Toronto Business   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1                            Garden District, Ryerson  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         The Beaches  43.676357 -79.293031  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                            Richmond, Adelaide, King  43.650571 -79.384568  
8                        Dufferin, Dovercourt Village  43.669005 -79.442259  
9                                  The Danforth  East  43.685347 -79.338106  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21                           Forest Hill North & West  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                                 North Toronto West  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                    

In [79]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="DT_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [80]:
# create map of Downtown Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [81]:
CLIENT_ID = 'E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL' # your Foursquare ID
CLIENT_SECRET = 'PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL
CLIENT_SECRET:PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ


In [82]:
toronto_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

In [83]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


In [84]:
radius = 500
limit = 100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborhood_latitude, neighborhood_longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL&client_secret=PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ&ll=43.6542599,-79.3606359&v=20180605&radius=500&limit=100'

In [85]:
# send GET request and get trending venues
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '610682e6e0f824558e08c5a9'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 46,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

In [86]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [87]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-87-2e7946891582>:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3          Impact Kitchen           Restaurant  43.656369 -79.356980
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [88]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

46 venues were returned by Foursquare.


However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Downtown Toronto. So let's slice the original dataframe and create a new dataframe of the Downtown Toronto data.


In [64]:
downtown_data = df_new[df_new['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_data

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M5B  Downtown Toronto   
2         M5C  Downtown Toronto   
3         M5E  Downtown Toronto   
4         M5G  Downtown Toronto   
5         M6G  Downtown Toronto   
6         M5H  Downtown Toronto   
7         M5J  Downtown Toronto   
8         M5K  Downtown Toronto   
9         M5L  Downtown Toronto   
10        M5S  Downtown Toronto   
11        M5T  Downtown Toronto   
12        M5V  Downtown Toronto   
13        M4W  Downtown Toronto   
14        M5W  Downtown Toronto   
15        M4X  Downtown Toronto   
16        M5X  Downtown Toronto   
17        M4Y  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1                            Garden District, Ryerson  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         Berczy Park  43.644771 -79.373306  
4                                  Central Bay Street  43.657952 -79.387383  
5                                            Christie  43.669542 -79.422564  
6                            Richmond, Adelaide, King  43.650571 -79.384568  
7   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
8            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
9                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
10                     University of Toronto, Harbord  43.662696 -79.400049  
11          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
12  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
13                                           Rosedale  43.679563 -79.377529  
14                                     Enclave of M5E  43.646435 -79.374846  
15                        St. James Town, Cabbagetown  43.667967 -79.367675  
16             First Canadian Place, Underground city  43.648429 -79.382280  
17                               Church and Wellesley  43.665860 -79.383160

Let's get the geographical coordinates of Downtown Toronto.


In [65]:
address = 'Downtown Toronto, Canada'

geolocator = Nominatim(user_agent="DT_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


As we did with all of Toronto, let's visualizae Downtown Toronto the neighborhoods in it.


In [66]:
# create map of Downtown Toronto using latitude and longitude values
map_downtown_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(downtown_data['Latitude'], downtown_data['Longitude'], downtown_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown_toronto)  
    
map_downtown_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.


In [67]:
CLIENT_ID = 'E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL' # your Foursquare ID
CLIENT_SECRET = 'PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL
CLIENT_SECRET:PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ


#### Let's explore the first neighborhood in our dataframe.


In [68]:
downtown_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

Get the neighborhood's latitude and longitude values.


In [69]:
neighborhood_latitude = downtown_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = downtown_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = downtown_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in Regent Park, Harbourfront within a radius of 500 meters.


First, let's create the GET request URL. Name your URL url.

In [24]:
radius = 500
limit = 100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborhood_latitude, neighborhood_longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL&client_secret=PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ&ll=43.6542599,-79.3606359&v=20180605&radius=500&limit=100'

Send the GET request and examine the resutls

In [25]:
# send GET request and get trending venues
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '6106792624aae1771b366189'},
 'response': {'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 46,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.653446723052674,
          'lng': -79.3620167174383}],
        'distance': 143,
       

we know that all the information is in the items key. Before we proceed, let's borrow the get_category_type function

In [26]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [27]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-27-2e7946891582>:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3          Impact Kitchen           Restaurant  43.656369 -79.356980
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [28]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

46 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Downtown Toronto.


First, let's create the GET request URL. Name your URL url.

#### Let's create a function to repeat the same process to all the neighborhoods in Downtown Toronto

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [30]:
downtown_venues = getNearbyVenues(names=downtown_data['Neighborhood'],
                                   latitudes=downtown_data['Latitude'],
                                   longitudes=downtown_data['Longitude']
                                  )

Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Enclave of M5E
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


#### Let's check the size of the resulting dataframe

In [31]:
print(downtown_venues.shape)
downtown_venues.head()

(1217, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

Let's check how many venues were returned for each neighborhood

In [32]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Enclave of M5E                                                         98   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              64   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
St. James Town                                                         89   
St. James Town, Cabbagetown                                            45   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Enclave of M5E                                                          98   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               64   
Regent Park, Harbourfront                                               46   
Richmond, Adelaide, King                                               100   
Rosedale                                                                 4   
St. James Town                                                          89   
St. James Town, Cabbagetown                                             45   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          35   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            57              57   
CN Tower, King and Spadina, Railway Lands, Harb...     17              17   
Central Bay Street                                     68              68   
Christie                                               16              16   
Church and Wellesley                                   78              78   
Commerce Court, Victoria Hotel                        100             100   
Enclave of M5E                                         98              98   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union Station, Toronto Islands     100             100   
Kensington Market, Chinato

#### Let's find out how many unique categories can be curated from all the returned venues


In [33]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 210 uniques categories.


## 3. Analyze Each Neighborhood


In [34]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  ...  Theater  Theme Restaurant  Toy / Game Store  Trail  \
0         0  ...        0                 0                 0      0   
1         0  ...        0                 0                 0      0   
2         0  ...        0                 0                 0      0   
3         0  ...        0                 0                 0      0   
4         0  ...        0                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  
0                      0         0          0  
1                      0         0          0  
2                      0         0          0  
3                      0         0          0  
4                      0         0          0  

[5 rows x 210 columns]

In [35]:
downtown_onehot.shape

(1217, 210)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [51]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped

Neighborhood  Yoga Studio   Airport  \
0                                         Berczy Park     0.000000  0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.058824   
2                                  Central Bay Street     0.014706  0.000000   
3                                            Christie     0.000000  0.000000   
4                                Church and Wellesley     0.025641  0.000000   
5                      Commerce Court, Victoria Hotel     0.010000  0.000000   
6                                      Enclave of M5E     0.010204  0.000000   
7              First Canadian Place, Underground city     0.000000  0.000000   
8                            Garden District, Ryerson     0.000000  0.000000   
9   Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000   
10          Kensington Market, Chinatown, Grange Park     0.000000  0.000000   
11                          Regent Park, Harbourfront     0.021739  0.000000   
12                           Richmond, Adelaide, King     0.000000  0.000000   
13                                           Rosedale     0.000000  0.000000   
14                                     St. James Town     0.000000  0.000000   
15                        St. James Town, Cabbagetown     0.000000  0.000000   
16           Toronto Dominion Centre, Design Exchange     0.000000  0.000000   
17                     University of Toronto, Harbord     0.028571  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.058824      0.058824        0.117647         0.176471   
2             0.000000      0.000000        0.000000         0.000000   
3             0.000000      0.000000        0.000000         0.000000   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0.000000         0.000000   

    Airport Terminal  American Restaurant  Antique Shop  ...   Theater  \
0           0.000000             0.000000      0.000000  ...  0.000000   
1           0.058824             0.000000      0.000000  ...  0.000000   
2           0.000000             0.000000      0.000000  ...  0.000000   
3           0.000000             0.000000      0.000000  ...  0.000000   
4           0.000000             0.012821      0.000000  ...  0.012821   
5           0.000000             0.030000      0.000000  ...  0.000000   
6           0.000000             0.010204      0.010204  ...  0.000000   
7           0.000000             0.030000      0.000000  ...  0.010000   
8           0.000000             0.000000      0.000000  ...  0.020000   
9           0.000000             0.000000      0.000000  ...  0.010000   
10          0.000000             0.000000      0.000000  ...  0.000000   
11          0.000000             0.000000      0.021739  ...  0.043478   
12          0.000000             0.020000      0.000000  ...  0.010000   
13          0.000000             0.000000      0.000000  ...  0.000000   
14          0.000000  

#### Let's confirm the new size


In [52]:
downtown_grouped.shape

(18, 210)

#### Let's print each neighborhood along with the top 5 most common venues

In [53]:
num_top_venues = 5

for hood in downtown_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_grouped[downtown_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0          Bakery  0.05
1     Coffee Shop  0.05
2    Cocktail Bar  0.05
3  Farmers Market  0.04
4             Pub  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
             venue  freq
0  Airport Service  0.18
1   Airport Lounge  0.12
2            Plane  0.06
3          Airport  0.06
4         Boutique  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.18
1                Café  0.06
2      Sandwich Place  0.06
3  Italian Restaurant  0.04
4        Burger Joint  0.03


----Christie----
                venue  freq
0       Grocery Store  0.25
1                Café  0.19
2                Park  0.12
3         Candy Store  0.06
4  Italian Restaurant  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.08
1  Japanese Restaurant  0.08
2     Sushi Restaurant  0.06
3           Restaurant  0.04
4        

#### Let's put that into a _pandas_ dataframe


First, let's write a function to sort the venues in descending order.

In [54]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [55]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park                Bakery   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Coffee Shop          Cocktail Bar        Farmers Market   
1        Airport Lounge                 Plane               Airport   
2                  Café        Sandwich Place    Italian Restaurant   
3                  Café                  Park           Candy Store   
4   Japanese Restaurant      Sushi Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Pub            Restaurant              Pharmacy   
1              Boutique         Boat or Ferry      Sculpture Garden   
2          Burger Joint           Salad Place       Bubble Tea Shop   
3    Italian Restaurant             Nightclub            Baby Store   
4               Gay Bar                  Café          Dance Studio   

  8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
0              Beer Bar           Cheese Shop        Seafood Restaurant  
1                   Bar           Coffee Shop           Harbor / Marina  
2            Restaurant                   Spa       Japanese Restaurant  
3           Coffee Shop    Athletics & Sports                Restaurant  
4  Fast Food Restaurant                 Hotel  Mediterranean Restaurant

## 4. Cluster Neighborhoods


Run k-means to cluster the neighborhood into 4 clusters.

In [56]:
# set number of clusters
kclusters = 4

downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 2, 1, 3, 1, 1, 1, 1, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [57]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = downtown_data

# merge downtown_grouped with downtown_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

downtown_merged.head() # check the last columns!

PostalCode           Borough               Neighborhood   Latitude  \
0        M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   
1        M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
2        M5C  Downtown Toronto             St. James Town  43.651494   
3        M5E  Downtown Toronto                Berczy Park  43.644771   
4        M5G  Downtown Toronto         Central Bay Street  43.657952   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               1           Coffee Shop                   Pub   
1 -79.378937               1           Coffee Shop        Clothing Store   
2 -79.375418               1           Coffee Shop                  Café   
3 -79.373306               1                Bakery           Coffee Shop   
4 -79.387383               1           Coffee Shop                  Café   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park                Bakery        Breakfast Spot   
1   Japanese Restaurant                 Hotel                  Café   
2            Restaurant                Bakery        Cosmetics Shop   
3          Cocktail Bar        Farmers Market                   Pub   
4        Sandwich Place    Italian Restaurant          Burger Joint   

  6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
0               Theater                  Café                 Restaurant   
1        Cosmetics Shop       Bubble Tea Shop  Middle Eastern Restaurant   
2          Cocktail Bar        Clothing Store             Farmers Market   
3            Restaurant              Pharmacy                   Beer Bar   
4           Salad Place       Bubble Tea Shop                 Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0        Farmers Market         Chocolate Shop  
1      Ramen Restaurant   Fast Food Restaurant  
2              Creperie       Department Store  
3           Cheese Shop     Seafood Restaurant  
4                   Spa    Japanese Restaurant

Finally, let's visualize the resulting clusters

In [58]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighborhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters


Now, examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, then assign a name to each cluster.

#### Cluster 1


In [59]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13            Playground                 Trail           Yoga Studio   

   5th Most Common Venue     6th Most Common Venue 7th Most Common Venue  \
13                Museum  Mediterranean Restaurant           Men's Store   

   8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
13    Mexican Restaurant  Middle Eastern Restaurant     Miscellaneous Shop

#### Cluster 2


In [60]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               1           Coffee Shop   
1   Downtown Toronto               1           Coffee Shop   
2   Downtown Toronto               1           Coffee Shop   
3   Downtown Toronto               1                Bakery   
4   Downtown Toronto               1           Coffee Shop   
6   Downtown Toronto               1           Coffee Shop   
7   Downtown Toronto               1           Coffee Shop   
8   Downtown Toronto               1           Coffee Shop   
9   Downtown Toronto               1           Coffee Shop   
10  Downtown Toronto               1                  Café   
11  Downtown Toronto               1                  Café   
14  Downtown Toronto               1           Coffee Shop   
15  Downtown Toronto               1                  Café   
16  Downtown Toronto               1           Coffee Shop   
17  Downtown Toronto               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                    Pub                  Park                 Bakery   
1         Clothing Store   Japanese Restaurant                  Hotel   
2                   Café            Restaurant                 Bakery   
3            Coffee Shop          Cocktail Bar         Farmers Market   
4                   Café        Sandwich Place     Italian Restaurant   
6         Clothing Store                  Café                  Hotel   
7               Aquarium                 Hotel                   Café   
8                  Hotel                  Café             Restaurant   
9             Restaurant                 Hotel                   Café   
10                Bakery   Japanese Restaurant              Bookstore   
11                   Bar           Coffee Shop  Vietnamese Restaurant   
14    Seafood Restaurant    Italian Restaurant                    Pub   
15           Coffee Shop           Pizza Place                   Park   
16                  Café                 Hotel                    Gym   
17   Japanese Restaurant      Sushi Restaurant             Restaurant   

            5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Breakfast Spot               Theater                  Café   
1                            Café        Cosmetics Shop       Bubble Tea Shop   
2                  Cosmetics Shop          Cocktail Bar        Clothing Store   
3                             Pub            Restaurant              Pharmacy   
4                    Burger Joint           Salad Place       Bubble Tea Shop   
6                      Restaurant       Thai Restaurant                   Bar   
7                      Restaurant               Brewery        Scenic Lookout   
8             Japanese Restaurant           Salad Place    Seafood Restaurant   
9                             Gym   Japanese Restaurant         Deli / Bodega   
10                 Sandwich Place                   Bar           Yoga Studio   
11  Vegetarian / Vegan Restaurant  Caribbean Restaurant           Gaming Cafe   
14                         Bakery          Cocktail Bar                  Café   
15                         Bakery    Italian Restaurant            Restaurant   
16            Japanese Restaurant            Restaurant                Bakery   
17                        Gay Bar                  Café          Dance Studio   

        8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
0                  Restaurant        Farmers Market            Chocolate Shop  
1   Middle Eastern Restaurant      Ramen Restaurant      Fast Food Restaurant  
2              Farmers Market              Creperie          Department Store  
3                    Beer Bar           Cheese Shop        Seafood Restaurant  
4                  Restaurant                   Spa       Japanese Restaurant  
6                         Gym            Steakhouse                 Bookst

#### Cluster 3


In [61]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12  Downtown Toronto               2       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12        Airport Lounge                 Plane               Airport   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12              Boutique         Boat or Ferry      Sculpture Garden   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12                   Bar           Coffee Shop        Harbor / Marina

#### Cluster 4


In [62]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5  Downtown Toronto               3         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5                  Café                  Park           Candy Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5    Italian Restaurant             Nightclub            Baby Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5           Coffee Shop    Athletics & Sports             Restaurant

#### Cluster 5

In [63]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Empty DataFrame
Columns: [Borough, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue, 8th Most Common Venue, 9th Most Common Venue, 10th Most Common Venue]
Index: []

end